text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,425 @@
+
+`cd /Users/poliorcetics/work/work-project && cargo build` produces artifacts directly in `/cargo-cache/debug/...`
+
+A subsequent `cargo build` in `perso-1` works with the same artifact, potentially having conflicting features for dependencies for example.
+
+A `cargo clean` deletes the entire `/cargo-cache` directory, for all projects at once.
+
+It's possible to produce invalid state in the target dir by having unrelated projects writing in the same place.
+
+It's not possible to have to projects building at once because Cargo locks its target directory during builds.
+
+#### With `CARGO_TARGET_DIR=/cargo-cache/{manifest-path-hash}`
+
+`cd /Users/poliorcetics/work/work-project && cargo build` produces artifacts in `/cargo-cache/<manifest-path-hash>/debug/...` (where `manifest-path-hash` is a directory or several chained directories unique to the workspace, with an unspecified naming scheme).
+
+A `cargo build` in `perso-1` produces new artifacts in `/cargo-cache/<manifest-path-hash>/debug/...`.
+
+A `cargo clean` only removed the `/cargo-cache/<manifest-path-hash>/` subdirectory, not all the artifacts for all other projects that are also in the cache.
+
+In this situation, it's much less likely for Cargo to produce invalid state without a `build.rs` deliberately writing outside its target directory.
+
+Two projects can be built in parallel without troubles.
+
+#### Absolute and relative paths
+
+`CARGO_TARGET_DIR` can be either a relative or absolute path, which makes sense since it's mostly intended for a single project, which can then work from its own position to configure the target directory, and that stays the case with templates.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+## Templating values of `CARGO_TARGET_DIR`
+
+Templating does not interfere with the resolution order of `CARGO_TARGET_DIR`. From less to most specific:
+
+- Through the `config.toml`:
+
+  ```toml
+  [build]
+  target-dir = "/absolute/path/to/cache/{manifest-path-hash}"
+  ```
+
+- Through the environment variable: `CARGO_TARGET_DIR="/absolute/path/to/cache/{manifest-path-hash}" cargo build`
+- Through the command line flag: `cargo build --target-dir "/absolute/path/to/cache/{manifest-path-hash}"`
+
+## Definition of `{manifest-path-hash}`
+
+In the example in the previous section, `{manifest-path-hash}` was replaced with a relative path. This relative path is computed from the full and canonicalized path to the manifest for the workspace `Cargo.toml` (or the `script.rs` file directly for cargo-scripts).
```

Comment:
This will be the very first templating supported in `.cargo/config.toml`. Is the syntax flexible enough to extend to other configuration like [`[build.rustflags]`](https://doc.rust-lang.org/nightly/cargo/reference/config.html#buildrustflags), [`[env]`](https://doc.rust-lang.org/nightly/cargo/reference/config.html#env). I can see it useful in other issues like

* https://github.com/rust-lang/cargo/issues/12193
* https://github.com/rust-lang/cargo/issues/7352

Reply:
I see no reason for it to be limited.

`{name}` is not used anywhere currently IIRC and the most likely places it can crop up unexpectedly are paths, where this RFC already discuss how unlikely that is.

I suppose anything where cargo shells out to another program can also have it come up, but I'm not sure we can even design for every possible `target.TTT.runner` in eternity

Reply:
Nice. Thanks for the reply. Maybe it is worth putting this in future posibilities in some way?

Reply:
Done in https://github.com/rust-lang/rfcs/pull/3371/commits/d549cb12889d5a2c4a4dd64b5a8a0e3ee7b49aa0 :) I tried noting possible future problems I could find in 10 minutes of thinking and why they're not limiting